### ISSUE TITLE
EM equivalent to Open3.popen3 (from Trac)

### ISSUE THREAD

**Issue by jakedouglas** [eventmachine/eventmachine#15]
# Moved from Trac

Reported by:    careo   Assigned to:  
Priority:   major   Milestone:  
Keywords:       Cc:  
Description ¶

EM.popen behaves the same as IO.popen, which fails to capture stderr.

It'd be nice if there was an EM.popen3, or if EM.popen could take a stderr handler, that would capture the stderr of the child process.

03/17/09 01:55:02 changed by careo ¶

A pure-Ruby version of the above (with a painfully-bad "api") that works in both os x and linux can be found here: http://gist.github.com/80296 .

in a nutshell, it does is:
1. open a new pipe
2. reopen stderr to the writable end
3. does the EM.popen as usual, passing in
4. fixes stderr, 

The confusing bit is that the stderr handler doesn't...handle... in linux if it uses receive_data, but works fine in os x. Everything works all around if it uses notify_readable however.
03/29/09 19:23:08 changed by tmm1 ¶

On Linux, calling recv() on a pipe was failing. Latest git uses read() instead.
